Move CSSSelectorParserContext to its own file

nt1m opened the pull request:
#### b7c822739450db84fb5f89f0969112db84a9c934
<pre>
Move CSSSelectorParserContext to its own file
<a href="https://bugs.webkit.org/show_bug.cgi?id=265220">https://bugs.webkit.org/show_bug.cgi?id=265220</a>
<a href="https://rdar.apple.com/118700294">rdar://118700294</a>

Reviewed by Jean-Yves Avenard.

Allow constructing CSSSelectorParserContext outside of CSSSelectorParser without including the entirety of CSSSelectorParser.h

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext): Deleted.
(WebCore::add): Deleted.
* Source/WebCore/css/parser/CSSSelectorParser.h:
(WebCore::CSSSelectorParserContextHash::hash): Deleted.
(WebCore::CSSSelectorParserContextHash::equal): Deleted.
(WTF::HashTraits&lt;WebCore::CSSSelectorParserContext&gt;::constructDeletedValue): Deleted.
(WTF::HashTraits&lt;WebCore::CSSSelectorParserContext&gt;::isDeletedValue): Deleted.
(WTF::HashTraits&lt;WebCore::CSSSelectorParserContext&gt;::emptyValue): Deleted.
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp: Added.
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h: Added.
(WebCore::CSSSelectorParserContextHash::hash):
(WebCore::CSSSelectorParserContextHash::equal):
(WTF::HashTraits&lt;WebCore::CSSSelectorParserContext&gt;::constructDeletedValue):
(WTF::HashTraits&lt;WebCore::CSSSelectorParserContext&gt;::isDeletedValue):
(WTF::HashTraits&lt;WebCore::CSSSelectorParserContext&gt;::emptyValue):

Canonical link: <a href="https://commits.webkit.org/271035@main">https://commits.webkit.org/271035@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3dd71ca8d6e8bb237aebcaa44147c9fc56d334d0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27171 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5787 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28407 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/29378 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24843 "Hash 3dd71ca8 for PR 20807 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27629 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7677 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3182 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/29378 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27433 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/49/builds/7677 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/14/builds/28407 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/29378 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/49/builds/7677 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/14/builds/28407 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/30013 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/7677 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/14/builds/28407 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/30013 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4133 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/51/builds/3182 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/30013 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5597 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/14/builds/28407 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4592 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3511 "Built successfully and passed tests") | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4511 "Hash 3dd71ca8 for PR 20807 does not build (failure)") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/css/parser/CSSSelectorParser.h:
```diff
@@ -34,9 +34,8 @@
 #include "CSSParserSelector.h"
 #include "CSSParserTokenRange.h"
 #include "CSSSelectorList.h"
+#include "CSSSelectorParserContext.h"
```
You are only using a reference to it ; so you can just forward declare it, no need to include the header.

I think header should be only used in the code if it can be, lower the risk of compilation time increasing over time like it so often does.

jyavenard commented on Source/WebCore/css/parser/CSSSelectorParser.h:
```diff
@@ -45,22 +44,6 @@ class CSSSelectorList;
 class StyleSheetContents;
```
forward declare it here.
struct CSSSelectorParserContext;

nt1m replied:
Unfortunately I can't because of the `const CSSSelectorParserContext m_context;` member, it will complain about incomplete type

jyavenard commented on Source/WebCore/css/parser/CSSSelectorParserContext.h:
```diff
@@ -0,0 +1,73 @@
+    CSSSelectorParserContext(const CSSParserContext&);
+    explicit CSSSelectorParserContext(const Document&);
+
+    friend bool operator==(const CSSSelectorParserContext&, const CSSSelectorParserContext&) = default;
```
existing code but I'm surprised in needs to be made friend, nor that this method definition is even needed.

nt1m replied:
Removed `friend`

nt1m committed:
Move CSSSelectorParserContext to its own file
https://bugs.webkit.org/show_bug.cgi?id=265220
rdar://118700294

Reviewed by Jean-Yves Avenard.

Allow constructing CSSSelectorParserContext outside of CSSSelectorParser without including the entirety of CSSSelectorParser.h

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext): Deleted.
(WebCore::add): Deleted.
* Source/WebCore/css/parser/CSSSelectorParser.h:
(WebCore::CSSSelectorParserContextHash::hash): Deleted.
(WebCore::CSSSelectorParserContextHash::equal): Deleted.
(WTF::HashTraits<WebCore::CSSSelectorParserContext>::constructDeletedValue): Deleted.
(WTF::HashTraits<WebCore::CSSSelectorParserContext>::isDeletedValue): Deleted.
(WTF::HashTraits<WebCore::CSSSelectorParserContext>::emptyValue): Deleted.
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp: Added.
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h: Added.
(WebCore::CSSSelectorParserContextHash::hash):
(WebCore::CSSSelectorParserContextHash::equal):
(WTF::HashTraits<WebCore::CSSSelectorParserContext>::constructDeletedValue):
(WTF::HashTraits<WebCore::CSSSelectorParserContext>::isDeletedValue):
(WTF::HashTraits<WebCore::CSSSelectorParserContext>::emptyValue):

Canonical link: https://commits.webkit.org/271035@main
